### PR TITLE
fix(scylla-bench-thread): Run sb read cmd with timestamp from sb write

### PR DESCRIPTION
**Pull request by aleksbykov**
Scylla-bench with workload=timeseries in read mode should use
start timestamp from scylla-bench command in write mode.

In case of fix Issue 4241 to scylla_bench_thread was added
file log reader which parse output from write command with
timeserieis workload and pass it to read command

## PR pre-checks (self review)

- [ ] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [ ] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

**Comment by aleksbykov** [comment]
Job is running: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/abykov/job/BYO-abykov-staging-3/19/execution/node/72/log/

File: sdcm/scylla_bench_thread.py [hunk 1]
```diff
@@ -56,6 +69,34 @@ def run(self):
...
+        self.sb_log_filename = sb_log_filename
+        self.parent_cluster = node.parent_cluster
+        self.sb_mode = sb_mode
+        self.sb_workload = sb_workload
+        self.pattern = re.compile(r"Start timestamp:\s+(\d+)")
+
+    def run(self):
+        if self.sb_mode != ScyllaBenchModes.WRITE.value and self.sb_workload != ScyllaBenchWorkloads.TIMESERIES.value:
+            self.stop()
+        while not self.stopped():
```
Comment (juliayakovlev) [inline sdcm/scylla_bench_thread.py:84]: Maybe need to set timeout here

Comment (aleksbykov) [reply sdcm/scylla_bench_thread.py:84]: I think it is not needed, because this file follower only need for timeseries write

File: sdcm/cluster.py [hunk 2]
```diff
@@ -4405,6 +4405,7 @@ def __init__(self, params):
         self.params = params
         self._gemini_version = None
         self._gemini_base_path = None
+        self.sb_write_timeserieis_ts = None
```
Comment (enaydanov) [inline sdcm/cluster.py:4408]: You have a typo here.

**Review by enaydanov — CHANGES_REQUESTED** [review]

File: test-cases/longevity/longevity-twcs-48h.yaml [hunk 3]
```diff
@@ -21,3 +21,5 @@ nemesis_during_prepare: false
 space_node_threshold: 64424
 
 user_prefix: 'longevity-twcs-48h'
+
+post_prepare_cql_cmds: "ALTER TABLE scylla_bench.test with gc_grace_seconds = 300 and default_time_to_live = 21600 and compaction = {'class':'TimeWindowCompactionStrategy', 'compaction_window_size': 60, 'compaction_window_unit': 'MINUTES'};"
```
Comment (roydahan) [inline test-cases/longevity/longevity-twcs-48h.yaml:25]: But we don't have prepare_workload here.
What would happen in this case? Do we even have the table?
If not, you need to prepare schema or run a prepare workload that just create the table with 1 partition or something like that.

File: test-cases/longevity/longevity-twcs-3h.yaml [hunk 4]
```diff
@@ -3,8 +3,8 @@ bench_run: true
...
 stress_read_cmd: [
-    "scylla-bench -workload=timeseries -mode=read -partition-count=20000 -concurrency=100 -replication-factor=3 -clustering-row-count=10000000 -clustering-row-size=200  -rows-per-request=100 -start-timestamp=SCT_TIME -write-rate 125 -duration=10m -distribution hnormal --connection-count 100 -duration=170m",
-    "scylla-bench -workload=timeseries -mode=read -partition-count=20000 -concurrency=100 -replication-factor=3 -clustering-row-count=10000000 -clustering-row-size=200  -rows-per-request=100 -start-timestamp=SCT_TIME -write-rate 125 -duration=10m -distribution uniform --connection-count 100 -duration=170m"
+    "scylla-bench -workload=timeseries -mode=read -partition-count=20000 -concurrency=100 -replication-factor=3 -clustering-row-count=10000000 -clustering-row-size=200  -rows-per-request=100 -start-timestamp=GET_WRITE_TIMESTAMP -write-rate 125 -duration=10m -distribution hnormal --connection-count 100 -duration=170m",
```
Comment (roydahan) [inline test-cases/longevity/longevity-twcs-3h.yaml:6]: We need to do the same also for "write_rate" it's also a calculated value.
Although here it's easier because it's not something that changes and I think that the formula is "max_rate" / "partition_count".

File: sdcm/scylla_bench_thread.py [hunk 5]
```diff
@@ -56,6 +69,34 @@ def run(self):
                         event.add_info(node=self.node, line=line, line_number=line_number).publish()
 
 
+class ScyllaBenchTimeSeriesWriteTimestampStartSearcher(FileFollowerThread):
```
Comment (dkropachev) [inline sdcm/scylla_bench_thread.py:72]: This is overkill, just provide `time.time()` to `-start-timestamp` on `write`, remember it and pick it up on read

**Review by dkropachev — CHANGES_REQUESTED** [review]

File: sdcm/cluster.py (hunk 2 above)
Comment (aleksbykov) [reply sdcm/cluster.py:4408]: fixed

File: test-cases/longevity/longevity-twcs-48h.yaml (hunk 3 above)
Comment (aleksbykov) [reply test-cases/longevity/longevity-twcs-48h.yaml:25]: TWCS tests are run with `longevity_twcs_test.py`. This test define the prepare command in code explicitly. you are right, it could be confusing. may be it is better to move the setting also to code inside method of test?

File: test-cases/longevity/longevity-twcs-3h.yaml (hunk 4 above)
Comment (aleksbykov) [reply test-cases/longevity/longevity-twcs-3h.yaml:6]: As Avi suggest in comment: https://github.com/scylladb/scylla/issues/8030#issuecomment-1005017695 the -write-rate could be counted as max-rate * expected delay = 50k * 10ms = 500
But according your formula write rate is already have valid value: max-rate / partition_cound = 50K / 400 = 125

I set medium from both results: 250.

File: sdcm/scylla_bench_thread.py [hunk 6]
```diff
@@ -106,6 +147,11 @@ def __init__(self, stress_cmd, loader_set, timeout, node_list=None, round_robin=
         self.results_futures = []
         self.shell_marker = generate_random_string(20)
         self.max_workers = 0
+        # Find stress mode:
+        #    "scylla-bench -workload=sequential -mode=write -replication-factor=3 -partition-count=100"
+        #    "scylla-bench -workload=uniform -mode=read -replication-factor=3 -partition-count=100"
+        self.sb_mode = re.search(r"-mode=(.+?) ", stress_cmd).group(1)
```
[the remark this answers is not in the document]
Comment (aleksbykov) [reply sdcm/scylla_bench_thread.py:153]: @dkropachev i check our scylla-bench commands, each has the -mode and -workload. So ideally the pattern will always found. otherwise scylla-bench will failed to start. I don't think it is need to set default value.
Could you suggest to provide how it is better to handle? log warning, raise an exception? or check on config stage?

File: sdcm/scylla_bench_thread.py (hunk 5 above)
Comment (aleksbykov) [reply sdcm/scylla_bench_thread.py:72]: Will it be custom timestamp for cell when it is written or it is the value for twcs cluster key?
@roydahan what do you think?

Comment (aleksbykov) [reply sdcm/scylla_bench_thread.py:72]: @dkropachev implement as you suggest.
running job to validate

File: sdcm/scylla_bench_thread.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (aleksbykov) [reply sdcm/scylla_bench_thread.py:153]: Add additional check to sct_config. So -mode and -workload alwasy be in command and search will not failed

File: sdcm/scylla_bench_thread.py (hunk 5 above)
Comment (roydahan) [reply sdcm/scylla_bench_thread.py:72]: @dkropachev are you sure?
If so, @aleksbykov I think this is already the behaviour.

File: test-cases/longevity/longevity-twcs-3h.yaml [hunk 7]
```diff
@@ -1,10 +1,10 @@
...
-    "scylla-bench -workload=timeseries -mode=read -partition-count=20000 -concurrency=100 -replication-factor=3 -clustering-row-count=10000000 -clustering-row-size=200  -rows-per-request=100 -start-timestamp=SCT_TIME -write-rate 125 -duration=10m -distribution uniform --connection-count 100 -duration=170m"
+    "scylla-bench -workload=timeseries -mode=read -partition-count=20000 -concurrency=100 -replication-factor=3 -clustering-row-count=10000000 -clustering-row-size=200  -rows-per-request=100 -start-timestamp=GET_WRITE_TIMESTAMP -write-rate 250 -distribution hnormal --connection-count 100 -duration=170m",
+    "scylla-bench -workload=timeseries -mode=read -partition-count=20000 -concurrency=100 -replication-factor=3 -clustering-row-count=10000000 -clustering-row-size=200  -rows-per-request=100 -start-timestamp=GET_WRITE_TIMESTAMP -write-rate 250 -distribution uniform --connection-count 100 -duration=170m"
```
Comment (roydahan) [inline test-cases/longevity/longevity-twcs-3h.yaml:7]: Why did you change the write rate to 250?
It's also should be taken like the timestamp, or calculate it manually from the command:
max_write_rate / partition-count
50000 / 400 = 125.

File: test-cases/longevity/longevity-twcs-48h.yaml (hunk 3 above)
Comment (roydahan) [reply test-cases/longevity/longevity-twcs-48h.yaml:25]: I don't mind it'll be here, as long as it's really running :)

File: test-cases/longevity/longevity-twcs-3h.yaml [hunk 8]
```diff
@@ -21,3 +21,5 @@ nemesis_during_prepare: false
 space_node_threshold: 64424
 
 user_prefix: 'longevity-twcs-3h'
+
+post_prepare_cql_cmds: "ALTER TABLE scylla_bench.test with gc_grace_seconds = 300;"
```
Comment (roydahan) [inline test-cases/longevity/longevity-twcs-3h.yaml:25]: Here we don't have TTL and timewindow size?

File: sdcm/scylla_bench_thread.py (hunk 5 above)
Comment (roydahan) [reply sdcm/scylla_bench_thread.py:72]: I prefer to keep it like this if it works well

Comment (aleksbykov) [reply sdcm/scylla_bench_thread.py]: Job is running succesfull with custom time. https://jenkins.scylladb.com/job/scylla-staging/job/abykov/job/longevity-twcs-test/2/consoleFull
According scylla-bench code, the starttime will be taken from -starttime if it tpassed,  if not it just take Time.Now(). so passing the timestamp to write and read mode doesn't require this overhead.

File: test-cases/longevity/longevity-twcs-3h.yaml (hunk 7 above)
Comment (aleksbykov) [reply test-cases/longevity/longevity-twcs-3h.yaml:7]: I explain it here: https://github.com/scylladb/scylla-cluster-tests/pull/4266#discussion_r781054091

Comment (aleksbykov) [reply test-cases/longevity/longevity-twcs-3h.yaml:7]: i can revert back to 125.

File: test-cases/longevity/longevity-twcs-3h.yaml (hunk 8 above)
Comment (aleksbykov) [reply test-cases/longevity/longevity-twcs-3h.yaml:25]: as we discussed, the default values are ok for 3h
timewindow size 1 minute. ttl = 900

File: sdcm/scylla_bench_thread.py (hunk 5 above)
Comment (aleksbykov) [reply sdcm/scylla_bench_thread.py]: @roydahan monitor is http://54.154.21.81:3000

File: test-cases/longevity/longevity-twcs-3h.yaml (hunk 8 above)
Comment (roydahan) [reply test-cases/longevity/longevity-twcs-3h.yaml:25]: ok.

File: test-cases/longevity/longevity-twcs-3h.yaml (hunk 4 above)
Comment (roydahan) [reply test-cases/longevity/longevity-twcs-3h.yaml]: AFAIU, The write-rate here is just a hint for the read workload so it won't run the reads too fast and reach to points where the write workload hasn't written yet.
It has nothing to do with what Avi asked. for that we set the parallelism and the max-rate.

Comment (roydahan) [reply test-cases/longevity/longevity-twcs-3h.yaml]: Also, I didn't calculated it, I got it from the write workload output, same as with the timestamp.

File: sdcm/scylla_bench_thread.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (soyacz) [reply sdcm/scylla_bench_thread.py:125]: can you make sb_mode be of `ScyllaBenchModes` type? then all the if statement could compare without `.value` extraction. This would add more sense to this `ScyllaBenchModes` enum.

**Comment by roydahan** [comment]
@aleksbykov is it ready from your point of view?

File: test-cases/longevity/longevity-twcs-3h.yaml (hunk 4 above)
Comment (aleksbykov) [reply test-cases/longevity/longevity-twcs-3h.yaml]: i added comments to config yaml. How it is correctly should be calculated.

File: sdcm/scylla_bench_thread.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (aleksbykov) [reply sdcm/scylla_bench_thread.py]: @soyacz done. Thanks

File: test-cases/longevity/longevity-twcs-3h.yaml (hunk 7 above)
Comment (aleksbykov) [reply test-cases/longevity/longevity-twcs-3h.yaml]: Set to 125 and added comment to yaml config

**Comment by aleksbykov** [comment]
> @aleksbykov is it ready from your point of view?

@roydahan i think it is ready.

File: sdcm/scylla_bench_thread.py [hunk 9]
```diff
@@ -131,25 +153,32 @@ def verify_results(self):
...
+            node.parent_cluster.sb_write_timeseries_ts = write_timestamp = time.time_ns()
+            LOGGER.debug("Set start-time: %s", write_timestamp)
+            stress_cmd = re.sub(r"SET_WRITE_TIMESTAMP", f"{write_timestamp}", stress_cmd)
+            LOGGER.debug("Replaced stress command: %s", stress_cmd)
+
+        elif self.sb_mode == ScyllaBenchModes.READ and self.sb_workload == ScyllaBenchWorkloads.TIMESERIES:
+            def check_write_timestamp():
+                return node.parent_cluster.sb_write_timeseries_ts
+
+            write_timestamp = wait_for(check_write_timestamp, step=5, timeout=30)
```
Comment (dkropachev) [inline sdcm/scylla_bench_thread.py:166]:
```suggestion
            write_timestamp = wait_for(lambda: node.parent_cluster.sb_write_timeseries_ts, 
                step=5, 
                timeout=30,
                text='Waiting for "scylla-bench -workload=timeseries -mode=write" been started, to pick up timestamp'
                )
```

**Review by dkropachev — CHANGES_REQUESTED** [review]

File: sdcm/scylla_bench_thread.py (hunk 9 above)
Comment (aleksbykov) [reply sdcm/scylla_bench_thread.py]: fixed

**Review by roydahan — APPROVED** [review]

**Comment by roydahan** [comment]
@dkropachev if you approve please merge.

**Review by dkropachev — APPROVED** [review]

**Review by enaydanov — APPROVED** [review]

**Comment by aleksbykov** [comment]
@dkropachev Could you merge and backport?

File: test-cases/longevity/longevity-twcs-48h.yaml [hunk 10]
```diff
@@ -1,10 +1,13 @@
...
-    "scylla-bench -workload=timeseries -mode=read -partition-count=20000 -concurrency=100 -replication-factor=3 -clustering-row-count=10000000 -clustering-row-size=200  -rows-per-request=100 -start-timestamp=SCT_TIME -write-rate 125 -duration=10m -distribution uniform --connection-count 100 -duration=2880m"
+    "scylla-bench -workload=timeseries -mode=read -partition-count=20000 -concurrency=100 -replication-factor=3 -clustering-row-count=10000000 -clustering-row-size=200  -rows-per-request=100 -start-timestamp=GET_WRITE_TIMESTAMP -write-rate 125 -distribution hnormal --connection-count 100 -duration=2880m",
+    "scylla-bench -workload=timeseries -mode=read -partition-count=20000 -concurrency=100 -replication-factor=3 -clustering-row-count=10000000 -clustering-row-size=200  -rows-per-request=100 -start-timestamp=GET_WRITE_TIMESTAMP -write-rate 125 -distribution uniform --connection-count 100 -duration=2880m"
```
Comment (roydahan) [inline test-cases/longevity/longevity-twcs-48h.yaml:10]: @aleksbykov Why is the partition-count different between the write and the read?
I don't remember if it's intentional or a mistake.
